### PR TITLE
Add cassandra option test (part 1)

### PR DESCRIPTION
**Pull request by kevindiu**
### Description:

This PR includes the cassandra_option test code, and some refactoring for writing the test code.
It is only the part of the option test. 
This PR is merging against the [test/internal/db/cassandra_option](https://github.com/vdaas/vald/tree/test/internal/db/cassandra_option) branch. 

### Environment:

- Go Version: 1.14.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.0

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/db/nosql/cassandra/cassandra.go [hunk 1]
```diff
@@ -93,36 +80,55 @@ type client struct {
...
+		connectTimeout           time.Duration
+		port                     int
+		keyspace                 string
+		numConns                 int
+		consistency              gocql.Consistency
+		compressor               gocql.Compressor
+		username                 string
+		password                 string
+		authProvider             func(h *gocql.HostInfo) (gocql.Authenticator, error)
+		retryPolicy              retryPolicy
```
Comment (hlts2) [inline internal/db/nosql/cassandra/cassandra.go:97]: Could you please give me a reason of this implementation?
there are no method of `retryPolicy ` and `reconnectionPolicy `, `poolConfig `, `hostFilter ` .
So in my opinion, I think inner struct is better.

Comment (kevindiu) [reply internal/db/nosql/cassandra/cassandra.go:97]: it is only for testing. I need to implement the custom comparer for these struct, like https://github.com/vdaas/vald/pull/638/files#diff-de534877b6350fd05c5419656dbb4242R42
, so if I don't define the struct here, the comparer cannot refer to this struct definition.

File: internal/db/nosql/cassandra/option_test.go [hunk 2]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
+				dur: "dummy",
+			},
+			want: want{
+				obj: &T{
+					timeout: time.Minute,
+				},
+			},
+		},
+		{
+			name: "set timeout success if the time is nil",
```
Comment (hlts2) [inline internal/db/nosql/cassandra/option_test.go:417]:
```suggestion
			name: "set timeout success if the time is empty",
```

File: internal/db/nosql/cassandra/option_test.go [hunk 3]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
 )
 
+var (
+	// Goroutine leak is detected by `fastime`, but it should be ignored in the test because it is an external package.
+	goleakIgnoreOptions = []goleak.Option{
+		goleak.IgnoreTopFunction("github.com/kpango/fastime.(*Fastime).StartTimerD.func1"),
+	}
+
+	// default comparator option for client
+	clientOpts = []comparator.Option{
```
Comment (hlts2) [inline internal/db/nosql/cassandra/option_test.go:40]: I think `clientOpts` is general naming.
If we want to add code to the implementation file (_.go), I think naming may same idea.
so you should change name.
for example
```
clientOpts -> comparatorClientOpts
```

**Review by hlts2 — CHANGES_REQUESTED** [review]

**Review by hlts2 — CHANGES_REQUESTED** [review]
And could you please add comments to `option.go`?:pray:

**Review by hlts2 — APPROVED** [review]
LGTM

File: internal/db/nosql/cassandra/option_test.go [hunk 4]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got := WithHosts(test.args.hosts...)
-			   obj := new(T)
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithHosts(test.args.hosts...)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:158]:
```suggestion
			obj := new(T)
			got(obj)
			if err := test.checkFunc(test.want, obj, got); err != nil {
```
No need to change template code more than necessary

File: internal/db/nosql/cassandra/option_test.go [hunk 5]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithDialer(test.args.der)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
+				tt.Errorf("error = %v", err)
+			}
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:235]: same

File: internal/db/nosql/cassandra/option_test.go [hunk 6]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithCQLVersion(test.args.version)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
+				tt.Errorf("error = %v", err)
+			}
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:298]: same

File: internal/db/nosql/cassandra/option_test.go [hunk 7]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithTimeout(test.args.dur)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
+				tt.Errorf("error = %v", err)
+			}
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:444]: same

File: internal/db/nosql/cassandra/option_test.go [hunk 8]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithProtoVersion(test.args.version)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
+				tt.Errorf("error = %v", err)
+			}
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:361]: same

File: internal/db/nosql/cassandra/option_test.go [hunk 9]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithConnectTimeout(test.args.dur)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
+				tt.Errorf("error = %v", err)
+			}
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:528]: same

File: internal/db/nosql/cassandra/option_test.go [hunk 10]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithPort(test.args.port)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
+				tt.Errorf("error = %v", err)
+			}
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:592]: same

File: internal/db/nosql/cassandra/option_test.go [hunk 11]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithKeyspace(test.args.keyspace)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
+				tt.Errorf("error = %v", err)
+			}
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:655]: same

File: internal/db/nosql/cassandra/option_test.go [hunk 12]
```diff
@@ -19,1026 +19,703 @@ package cassandra
...
-			   got(obj)
-			   if err := test.checkFunc(tt.want, obj); err != nil {
-			       tt.Errorf("error = %v", err)
-			   }
-			*/
+			got := WithNumConns(test.args.numConns)
+			obj := new(T)
+			if err := test.checkFunc(test.want, obj, got(obj)); err != nil {
+				tt.Errorf("error = %v", err)
+			}
```
Comment (vankichi) [inline internal/db/nosql/cassandra/option_test.go:718]: same

File: internal/db/nosql/cassandra/option_test.go (hunk 4 above)
Comment (kevindiu) [reply internal/db/nosql/cassandra/option_test.go:158]: It is from template code for `option that return error` case. (line 133)

Comment (vankichi) [reply internal/db/nosql/cassandra/option_test.go:158]: oh, sorry

**Comment by vankichi** [comment]
🤔

**Comment by vankichi** [comment]
Your branch was created wrong base branch

**Pull request closed without merging**